optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/util/ConcurrentCache.java

```diff
@@ -0,0 +1,41 @@
+package org.optaplanner.core.impl.domain.solution.util;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.Function;
+
+/**
+ * Thread-safe basic cache implementation.
+ *
+ * @param <K> the type of keys for the cache
+ * @param <V> the type of values stored in the cache
+ * */
+public final class ConcurrentCache<K, V> implements Cache<K, V> {
+
+    private final ConcurrentMap<K, V> cache = new ConcurrentHashMap<>();
+
+    /**
+     * {@link ConcurrentHashMap#computeIfAbsent(Object, Function)} should not be used to directly retrieve cached values
+     * as it incorporates locking even when the value is present and doesn't have to be computed. Thus, the value is
+     * retrieved by {@link ConcurrentHashMap#get(Object)} and iff the value is null,
```

Comment:
Typo.

Reply:
if "iff" is the abbreviation of "if and only if" let's write it out the long way, not everyone is a math geek ;)

Reply:
Didn't even cross my mind that it could actually mean something. Well, I guess "not everyone is a math geek." :-)

Reply:
I don't consider myself a math geek and all professors who used to give me lessons in Math would support my self-assessment :)

But yes, will write it in the full form.